src/app/clusters/service-area-server/service-area-server.h

```diff
@@ -0,0 +1,751 @@
+     * @param aLocationId the locationId to use
+     * @param aEstimatedTime total operational time for thus location
+     * @return true if progress element is found and operational time is set
+     * @note caller is responsible for change notification
+    */
+    bool SetProgressEstimatedTime(uint32_t aLocationId, const DataModel::Nullable<uint32_t> & aEstimatedTime);
+
+    /**
+     * @brief remove any progress elements which no longer reference a supported location (locationId not in supported locations)
+     * @return true if any progress elements were removed.
+     * @note caller is responsible for change notification
+     */
+    bool PruneProgressElements();
+
+        /**
+     * @brief remove a location from the progress list
+     * @return true if the location was removed from the progress list.
+     * @note caller is responsible for change notification
+     */
+    bool DeleteProgressElement(uint32_t aLocationId);
+
+    /**
+     * @brief Clear the progress list
+     * @note caller is responsible for change notification
+     * @return the Progress list changed (i.e. it previously had entries)
+
+     */
+    bool ClearProgressList();
+
+
+    //*************************************************************************
+    // Feature Map attribute
+
+    /**
+     * @brief Check if a feature is supported
+     * @param feature the feature enum
+     * @return true if the feature is supported
+     * @note the Service Area features are set at startup and are read-only to both device and client
+     */
+
+    bool HasFeature(ServiceArea::Feature feature) const;
+};
+
+
+/**
+ * Location::Delegate Defines methods for implementing application-specific
+ * logic for the Location Cluster.
+ */
+class Delegate : public Uncopyable
+{
+public:
+    Delegate() = default;
+    virtual ~Delegate() = default;
+
+private:
+
+    friend class Instance;
+    Instance * mInstance = nullptr;
+
+    /**
+     * @brief This method is used by the SDK to set the instance pointer. This is done during the instantiation of a Instance object.
+     * @param aInstance A pointer to the Instance object related to this delegate object.
+     */
+    void SetInstance(Instance * aInstance) { mInstance = aInstance; }
+
+
+    //*************************************************************************
+    // volatile attribute support
+
+    /**
+     * @brief give the device a chance to update current location
+     * @param prevCurrentLocation "current" current location value - used to determine if update notification is triggered
+     * @note this is called by the instance function GetCurrentLocation() to ensure the latest value is available
+     * @note default implementation can be used if current location is non-volatile (always expected to be up-to-date)
+    */
+    virtual void HandleVolatileCurrentLocation(const DataModel::Nullable<uint32_t> prevCurrentLocation) = 0;
```

Comment:
I'm really uncertain about these, they provide flexibility for the application/device implementer, but the more MATTER code and specs I looked at, I think they may not fit well with the presumed design principles.

These in theory allow the server code to query the application/device whenever one of these attributes is read,  a "pull model".

I haven't done a deep dive yet, but it seems like it's preferred to either have the device publish updates to the application (push model), or have the application poll the device (at 1 second ticks?) - poll model.

At this point I'm 50/50, I figured it was better to leave them in and see what people thought.

Reply:
Interesting topic. My first impulse would be to say that the app should update the server as soon as possible, to ensure the server always has a consistent state/view of the device's state. 

I'd be curious to see what Boris and Tennessee say, if they'll comment about it.

Reply:
Yes, if the app keeps the server up-to-date, then these functions are not needed.  That's my later thought that I'm leaning toward.   I'll be digging into adding the service area to the irobot app now, that should give me a better perspective on this.